pkg/sentry/fsimpl/erofs/regular_file.go

```diff
@@ -0,0 +1,247 @@
+	d.mapsMu.Lock()
+	mapped := d.mappings.AddMapping(ms, ar, offset, writable)
+	for _, r := range mapped {
+		d.pf.hostFileMapper.IncRefOn(r)
+	}
+	d.mapsMu.Unlock()
+	return nil
+}
+
+// RemoveMapping implements memmap.Mappable.RemoveMapping.
+func (d *dentry) RemoveMapping(ctx context.Context, ms memmap.MappingSpace, ar hostarch.AddrRange, offset uint64, writable bool) {
+	d.mapsMu.Lock()
+	unmapped := d.mappings.RemoveMapping(ms, ar, offset, writable)
+	for _, r := range unmapped {
+		d.pf.hostFileMapper.DecRefOn(r)
+	}
+	d.mapsMu.Unlock()
+}
+
+// CopyMapping implements memmap.Mappable.CopyMapping.
+func (d *dentry) CopyMapping(ctx context.Context, ms memmap.MappingSpace, srcAR, dstAR hostarch.AddrRange, offset uint64, writable bool) error {
+	return d.AddMapping(ctx, ms, dstAR, offset, writable)
+}
+
+// Translate implements memmap.Mappable.Translate.
+func (d *dentry) Translate(ctx context.Context, required, optional memmap.MappableRange, at hostarch.AccessType) ([]memmap.Translation, error) {
+	pgend, _ := hostarch.PageRoundUp(d.inode.Size())
+	if required.End > pgend {
+		if required.Start >= pgend {
+			return nil, &memmap.BusError{io.EOF}
+		}
+		required.End = pgend
+	}
+	if optional.End > pgend {
+		optional.End = pgend
+	}
+	offset, err := d.inode.DataOffset()
+	if err != nil {
+		return nil, &memmap.BusError{err}
+	}
+	mr := optional
+	return []memmap.Translation{
+		{
+			Source: mr,
+			File:   &d.pf,
+			Offset: mr.Start + offset,
+			Perms:  hostarch.AnyAccess,
+		},
+	}, nil
+}
+
+// InvalidateUnsavable implements memmap.Mappable.InvalidateUnsavable.
+func (d *dentry) InvalidateUnsavable(ctx context.Context) error {
+	d.mapsMu.Lock()
+	d.mappings.InvalidateAll(memmap.InvalidateOpts{})
+	d.mapsMu.Unlock()
+	return nil
+}
+
+// +stateify savable
+type dentryPlatformFile struct {
+	*dentry
+
+	// fdRefsMu protects fdRefs.
+	fdRefsMu sync.Mutex `state:"nosave"`
+
+	// fdRefs counts references on memmap.File offsets.
+	// +checklocks:fdRefsMu
+	fdRefs fsutil.FrameRefSet
+
+	// hostFileMapper caches mappings of the file represented by this dentry.
+	hostFileMapper fsutil.HostFileMapper
+
+	// hostFileMapperInitOnce is used to lazily initialize hostFileMapper.
+	hostFileMapperInitOnce sync.Once `state:"nosave"`
+}
+
+// IncRef implements memmap.File.IncRef.
+func (d *dentryPlatformFile) IncRef(fr memmap.FileRange, memCgID uint32) {
+	d.fdRefsMu.Lock()
+	d.fdRefs.IncRefAndAccount(fr, memCgID)
+	d.fdRefsMu.Unlock()
+}
+
+// DecRef implements memmap.File.DecRef.
+func (d *dentryPlatformFile) DecRef(fr memmap.FileRange) {
+	d.fdRefsMu.Lock()
+	d.fdRefs.DecRefAndAccount(fr)
+	d.fdRefsMu.Unlock()
+}
+
+// MapInternal implements memmap.File.MapInternal.
+func (d *dentryPlatformFile) MapInternal(fr memmap.FileRange, at hostarch.AccessType) (safemem.BlockSeq, error) {
+	return d.hostFileMapper.MapInternal(fr, d.inode.FD(), at.Write)
```

Comment:
1. The base image is read-only, right? So pass `at.Read`? We don't want to modify the filesystem image in any way.
2. Can we just reuse `erofs.Image.bytes`? We have already mmap(2)-ed the entire filesystem image into sentry address space. Can't we just construct `safemem.BlockSeq` from that? `d.hostFileMapper.MapInternal()` goes ahead and mmap(2)s the whole file again. This costs more host VMAs. @nixprime am I suggesting something wrong? IDK what `(*dentryPlatformFile).FD()` should return in that case.

Reply:
> The base image is read-only, right? So pass at.Read? We don't want to modify the filesystem image in any way.

The third parameter of `d.hostFileMapper.MapInternal(`) is `write bool`, and `at.Write` indicates whether the write permission is requested by this map. That’s why we need to pass `at.Write` instead of `at.Read`. And if the write permission  is requested, the `mmap()` in `forEachMappingBlockLocked()` in `d.hostFileMapper.MapInternal()` will fail as the image fd is readonly. That's why it works. I will add a check in `MapInternal()` in `erofs.go` to make it more explicit.

https://github.com/google/gvisor/blob/f8f2bb6d21ba55b13d6356cb78d7039b2b0dfe51/pkg/sentry/fsutil/host_file_mapper.go#L148

> Can we just reuse erofs.Image.bytes? We have already mmap(2)-ed the entire filesystem image into sentry address space. Can't we just construct safemem.BlockSeq from that? d.hostFileMapper.MapInternal() goes ahead and mmap(2)s the whole file again. This costs more host VMAs. @nixprime am I suggesting something wrong? IDK what (*dentryPlatformFile).FD() should return in that case.

I had some similar thoughts. It might require some refactors in mm (e.g. for the `(*dentryPlatformFile).FD()` related parts). I prefer to do it in the follow up PRs.

Reply:
> The third parameter of d.hostFileMapper.MapInternal() is write bool, and at.Write indicates whether the write permission is requested by this map.

Ah right, my bad. I guess what I was getting to was that we should deny `at.Write`. But like you said, since the FD is read-only, that should be fine, it will automatically fail. Optionally, you could add a more explicit check.

> It might require some refactors in mm (e.g. for the (*dentryPlatformFile).FD() related parts). I prefer to do it in the follow up PRs.

Makes sense. Lets leave a comment about this over here for now and we can revisit it later.